sdks/java/harness/src/main/java/org/apache/beam/fn/harness/SplittablePairWithRestrictionDoFnRunner.java

```diff
@@ -0,0 +1,384 @@
+    @SuppressWarnings("rawtypes") // cannot do this multi-level cast without rawtypes
+    FnDataReceiver<WindowedValue<KV<InputT, KV<RestrictionT, WatermarkEstimatorStateT>>>>
+        mainOutputConsumer =
+            (FnDataReceiver)
+                getPCollectionConsumer.apply(pTransform.getOutputsMap().get(mainOutputTag.getId()));
+    this.mainOutputConsumer = mainOutputConsumer;
+
+    // Main input
+    String mainInputTag =
+        Iterables.getOnlyElement(
+            Sets.difference(
+                pTransform.getInputsMap().keySet(), parDoPayload.getSideInputsMap().keySet()));
+    String mainInputName = pTransform.getInputsOrThrow(mainInputTag);
+    PCollection mainInput =
+        components.getPcollectionsMap().get(pTransform.getInputsOrThrow(mainInputTag));
+
+    // Side inputs
+    this.sideInputMapping = ParDoTranslation.getSideInputMapping(parDoPayload);
+    @SuppressWarnings("rawtypes") // passed to FnApiStateAccessor which uses rawtypes
+    ImmutableMap.Builder<TupleTag<?>, SideInputSpec> tagToSideInputSpecMapBuilder =
+        ImmutableMap.builder();
+    for (Map.Entry<String, RunnerApi.SideInput> entry :
+        parDoPayload.getSideInputsMap().entrySet()) {
+      String sideInputTag = entry.getKey();
+      RunnerApi.SideInput sideInput = entry.getValue();
+      PCollection sideInputPCollection =
+          components.getPcollectionsMap().get(pTransform.getInputsOrThrow(sideInputTag));
+      WindowingStrategy<?, ?> sideInputWindowingStrategy =
+          rehydratedComponents.getWindowingStrategy(sideInputPCollection.getWindowingStrategyId());
+      tagToSideInputSpecMapBuilder.put(
+          new TupleTag<>(entry.getKey()),
+          SideInputSpec.create(
+              sideInput.getAccessPattern().getUrn(),
+              rehydratedComponents.getCoder(sideInputPCollection.getCoderId()),
+              (Coder<WindowT>) sideInputWindowingStrategy.getWindowFn().windowCoder(),
+              PCollectionViewTranslation.viewFnFromProto(entry.getValue().getViewFn()),
+              (WindowMappingFn<WindowT>)
+                  PCollectionViewTranslation.windowMappingFnFromProto(
+                      entry.getValue().getWindowMappingFn())));
+    }
+    @SuppressWarnings("rawtypes") // passed to FnApiStateAccessor which uses rawtypes
+    Map<TupleTag<?>, SideInputSpec> tagToSideInputSpecMap = tagToSideInputSpecMapBuilder.build();
+
+    // Register processing methods
+    if (doFnSignature.getInitialRestriction().observesWindow() || !sideInputMapping.isEmpty()) {
+      addPCollectionConsumer.accept(
+          mainInputName, this::processElementForWindowObservingPairWithRestriction);
+      this.mutableArgumentProvider = new WindowObservingProcessBundleContext();
+    } else {
+      addPCollectionConsumer.accept(mainInputName, this::processElementForPairWithRestriction);
+      this.mutableArgumentProvider = new NonWindowObservingProcessBundleContext();
+    }
+    addTearDownFunction.accept(this::tearDown);
```

Comment:
Not directly related to this PR, but notice that it does something for DoFn.teardown here, but there is a long standing issue DoFn.teardown isn't honored in Dataflow runner (both legacy and v2): go/beam-issue/18592 . Wondering if there is any gap in SDK harness

Reply:
Good point. I'll look closer as I factor this out. It seems there are many copies of the DoFn used for different things, and they'd all have their teardown registered separately too.